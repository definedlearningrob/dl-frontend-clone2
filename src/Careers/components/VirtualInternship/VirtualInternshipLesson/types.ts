import { TLesson as TStudentLesson } from '@dc/graphql/student/queries/virtualInternshipLesson';
import { TLesson as TUserLesson } from '@dc/graphql/user/queries/virtualInternshipLesson';

export type TLesson = TStudentLesson | TUserLesson;
export type TCheckInQuestion = TLesson['checkInQuestions'][number];
export type TCheckInGroup = TLesson['checkInGroups'][number];
