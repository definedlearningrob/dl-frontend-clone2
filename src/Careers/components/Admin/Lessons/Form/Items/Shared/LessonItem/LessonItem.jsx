import PropTypes from 'prop-types';

import { ReactComponent as DragHandle } from '@dc/svg/temp/drag_handle.svg';
import { ReactComponent as Remove } from '@dc/svg/remove.svg';

import SharedIcon from '@shared/components/Icon/Icon';
import DeprecatedIconButton from '@shared/components/DeprecatedIconButton/DeprecatedIconButton';
import { ReactComponent as Search } from '@shared/svg/search.svg';

AdminLessonsFormItemsSharedLessonItem.propTypes = {
  identifier: PropTypes.string,
  label: PropTypes.string,
  onMore: PropTypes.func,
  onRemove: PropTypes.func,
  testPrefix: PropTypes.string,
};

function AdminLessonsFormItemsSharedLessonItem({
  identifier,
  onRemove,
  onMore,
  label,
  testPrefix,
}) {
  return (
    <>
      <SharedIcon className='lessons__list-item__drag-icon' icon={<DragHandle />} size='sm' />
      <div className='lessons__list-item__type-badge'>{label}</div>
      <span className='lessons__list-item__title'>{identifier}</span>
      <div className='lessons__list-item__action-buttons'>
        <DeprecatedIconButton
          className='lessons__list-item__details-button'
          data-testid='lesson-item-more-info'
          icon={<Search />}
          size='sm'
          onClick={onMore}
        />
        <DeprecatedIconButton
          className='lessons__list-item__remove-button'
          data-testid={`${testPrefix}-remove`}
          icon={<Remove />}
          size='sm'
          onClick={onRemove}
        />
      </div>
    </>
  );
}

export default AdminLessonsFormItemsSharedLessonItem;
