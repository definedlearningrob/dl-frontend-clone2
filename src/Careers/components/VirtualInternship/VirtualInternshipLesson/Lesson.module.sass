@import '@shared/stylesheets/config.sass'

.contentWrapper
  display: grid
  grid-template-columns: 2fr 1fr
  gap: $m-base

  +media-breakpoint-up(xxxl)
    gap: $m-md

.resourceName
  color: $color-font-secondary
  font-size: $font-size-xs
  line-height: $line-height-lg
  letter-spacing: 0.01em
  margin-bottom: $m-xs

  +media-breakpoint-up(xxxl)
    font-size: $font-size-sm

.lessonName
  font-size: $font-size-base
  margin-bottom: $m-sm

  +media-breakpoint-up(xxxl)
    margin-bottom: $m-base

.lessonItems
  +flex-container(column)
  gap: $m-base

  +media-breakpoint-up(xxxl)
    gap: $m-md

.presentationsWrapper
  +flex-container(column)
  gap: $m-base
  margin-bottom: $m-base

  +media-breakpoint-up(xxxl)
    gap: $m-md
    margin-bottom: $m-md

.progressWrapper
  grid-column-start: 2
