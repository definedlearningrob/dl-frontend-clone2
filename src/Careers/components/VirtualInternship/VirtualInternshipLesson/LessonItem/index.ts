export * from './PresentationCard/PresentationCard';
export * from './AssignmentCard/AssignmentCard';
export * from './AttachmentCard/AttachmentCard';
export * from './ResearchLinksCard/ResearchLinksCard';
export * from './TextCard/TextCard';
export * from './VideoCard/VideoCard';
export * from './VocabularyCard/VocabularyCard';
