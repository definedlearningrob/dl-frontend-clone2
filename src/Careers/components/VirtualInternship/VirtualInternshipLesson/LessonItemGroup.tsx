import {
  TAssignment,
  TAttachment,
  TResearchLink,
  TText,
  TVideo,
  TVocabulary,
} from '@dc/components/Student/Lesson/types';

import {
  AssignmentCard,
  AttachmentCard,
  ResearchLinksCard,
  TextCard,
  VideoCard,
  VocabularyCard,
} from './LessonItem';

export type LessonItemGroup = (
  | {
      type: TVocabulary['__typename'];
      items: TVocabulary[];
    }
  | {
      type: TText['__typename'];
      items: TText[];
    }
  | {
      type: TAttachment['__typename'];
      items: TAttachment[];
    }
  | {
      type: TVideo['__typename'];
      items: TVideo[];
    }
  | {
      type: TAssignment['__typename'];
      items: TAssignment[];
    }
  | {
      type: TResearchLink['__typename'];
      items: TResearchLink[];
    }
) & {
  id: string;
};

type Props = {
  lessonItemGroup: LessonItemGroup;
  virtualInternshipId: string;
  isPreviewOnly?: boolean;
};

export const LessonItemGroup = ({ lessonItemGroup, virtualInternshipId, isPreviewOnly }: Props) => {
  switch (lessonItemGroup.type) {
    case 'Vocabulary':
      return <VocabularyCard id={lessonItemGroup.id} vocabularies={lessonItemGroup.items} />;
    case 'Text':
      return (
        <>
          {lessonItemGroup.items.map((text) => (
            <TextCard key={text.id} text={text} />
          ))}
        </>
      );
    case 'Assignment':
      return (
        <>
          {lessonItemGroup.items.map((assignment) => (
            <AssignmentCard
              key={assignment.id}
              assignment={assignment}
              isPreviewOnly={isPreviewOnly}
              virtualInternshipId={virtualInternshipId}
            />
          ))}
        </>
      );
    case 'Attachment':
      return (
        <>
          {lessonItemGroup.items.map((attachment) => (
            <AttachmentCard key={attachment.id} attachment={attachment} />
          ))}
        </>
      );
    case 'Video':
      return (
        <>
          {lessonItemGroup.items.map((video) => (
            <VideoCard key={video.id} video={video} />
          ))}
        </>
      );
    case 'ResearchLink':
      return <ResearchLinksCard id={lessonItemGroup.id} researchLinks={lessonItemGroup.items} />;
    default:
      return null;
  }
};
