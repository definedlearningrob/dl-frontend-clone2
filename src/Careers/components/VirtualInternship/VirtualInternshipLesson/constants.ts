export const GROUPABLE_ITEMS = ['ResearchLink', 'Vocabulary'];
export const COMPLETABLE_ITEMS = ['Assignment', 'CheckInQuestion', 'CheckInGroup'];
