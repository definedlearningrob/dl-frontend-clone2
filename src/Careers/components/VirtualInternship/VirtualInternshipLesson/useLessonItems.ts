import { sortBy } from 'lodash-es';
import { useMemo } from 'react';

import {
  TAssignment,
  TAttachment,
  TResearchLink,
  TText,
  TVideo,
  TVocabulary,
} from '@dc/components/Student/Lesson/types';
import type { LessonItemGroup } from '@dc/components/VirtualInternship/VirtualInternshipLesson/LessonItemGroup';

import { TLesson } from './types';

type LessonItem = TVocabulary | TAttachment | TAssignment | TText | TVideo | TResearchLink;
type StandaloneGroupType = Exclude<
  LessonItem['__typename'],
  TVocabulary['__typename'] | TResearchLink['__typename']
>;

type Props = {
  lesson: TLesson;
};

const standaloneGroups: StandaloneGroupType[] = ['Assignment', 'Attachment', 'Text', 'Video'];

export const useLessonItems = ({ lesson }: Props) => {
  const lessonItemGroups = useMemo(() => {
    const lessonItems = [
      lesson.vocabularies,
      lesson.texts,
      lesson.attachments,
      lesson.videos,
      lesson.assignments,
      lesson.researchLinks,
    ]
      .flat()
      .sort((a, b) => a.step - b.step);

    return lessonItems.reduce((acc, item) => {
      const previousGroup = acc.at(-1);

      const shouldCreateNewGroup =
        !previousGroup ||
        previousGroup.type !== item.__typename ||
        (standaloneGroups as string[]).includes(item.__typename);

      if (shouldCreateNewGroup) {
        acc.push({
          type: item.__typename,
          items: [item],
          id: item.id,
        } as LessonItemGroup);
      } else {
        (previousGroup.items as unknown[]).push(item);
      }

      return acc;
    }, [] as LessonItemGroup[]);
  }, [lesson]);

  const checkInItems = sortBy([...lesson.checkInGroups, ...lesson.checkInQuestions], 'step');
  const externalPresentations = sortBy(lesson.externalPresentations, 'step');

  return { lessonItemGroups, checkInItems, externalPresentations };
};
