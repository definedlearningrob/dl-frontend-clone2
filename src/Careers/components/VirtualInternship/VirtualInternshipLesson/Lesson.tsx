import { useEffect } from 'react';

import { LESSON_TYPES } from '@dc/resources/constants';
import { PresentationCard } from '@dc/components/VirtualInternship/VirtualInternshipLesson/LessonItem';
import { LessonItemGroup } from '@dc/components/VirtualInternship/VirtualInternshipLesson/LessonItemGroup';
import { LessonCheckInItems } from '@dc/components/VirtualInternship/VirtualInternshipLesson/LessonCheckInItems/LessonCheckInItems';
import { LessonTableOfContent } from '@dc/components/VirtualInternship/VirtualInternshipLesson/LessonTableOfContent/LessonTableOfContent';
import { CareerReviewSurvey } from '@dc/components/Student/Lesson/Survey/CareerReviewSurvey';

import useQueryParams from '@shared/hooks/useQueryParams';
import { cx } from '@shared/utils/cx';

import styles from './Lesson.module.sass';
import { useLessonItems } from './useLessonItems';
import { TLesson } from './types';

type Props = {
  resourceName: string;
  resourceId: string;
  lesson: TLesson;
  isPreviewOnly?: boolean;
};

export const Lesson = ({ lesson, resourceName, resourceId, isPreviewOnly }: Props) => {
  const {
    params: { section },
  } = useQueryParams<{ section?: string }>();

  const lessonItems = useLessonItems({ lesson });

  const isCareerReviewSurvey = lesson.type.toUpperCase() === LESSON_TYPES.CAREER_REVIEW_SURVEY;

  useEffect(() => {
    const lessonItem = section && document.getElementById(section);

    if (lessonItem) {
      const appHeaderHeight = 48;

      window.scrollTo({
        top: lessonItem.offsetTop - appHeaderHeight,
        behavior: 'smooth',
      });
    }
  }, [section]);

  // TODO: handle extension fields for Dig deeper into career lesson type

  return (
    <div>
      {!isCareerReviewSurvey && (
        <>
          <p className={styles.resourceName}>{resourceName}</p>
          <h4 className={styles.lessonName}>{lesson.name}</h4>
        </>
      )}
      {lesson.hasPresentation && (
        <div className={styles.presentationsWrapper}>
          {lessonItems.externalPresentations.map((presentation) => (
            <PresentationCard key={presentation.id} presentation={presentation} />
          ))}
        </div>
      )}
      <div className={cx({ 'grid grid-cols-[2fr_1fr] gap-base xxxl:gap-md': !isPreviewOnly })}>
        <div className={styles.lessonItems}>
          {lessonItems.lessonItemGroups.map((lessonItemGroup, index) => (
            <LessonItemGroup
              key={index}
              isPreviewOnly={isPreviewOnly}
              lessonItemGroup={lessonItemGroup}
              virtualInternshipId={resourceId}
            />
          ))}
          <LessonCheckInItems checkInItems={lessonItems.checkInItems} />
          {isCareerReviewSurvey && (
            <CareerReviewSurvey
              careerReviewSurvey={lesson.careerReviewSurvey}
              courseName={resourceName}
              previewOnly={isPreviewOnly}
              resourceId={resourceId}
            />
          )}
        </div>
        {!isPreviewOnly && (
          <div className={styles.progressWrapper}>
            <LessonTableOfContent />
          </div>
        )}
      </div>
    </div>
  );
};
