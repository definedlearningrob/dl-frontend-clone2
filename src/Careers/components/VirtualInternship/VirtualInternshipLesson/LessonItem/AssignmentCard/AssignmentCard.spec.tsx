import { MockedProvider } from '@apollo/client/testing';
import { Route } from 'react-router-dom';
import { fireEvent, screen, waitFor, within } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { renderWithRouterAndReduxProvider } from '@dc/utils/test';
import { AssignmentCard } from '@dc/components/VirtualInternship/VirtualInternshipLesson/LessonItem';
import deleteAssignmentSubmissionFileMutation from '@dc/graphql/student/mutations/deleteAssignmentSubmissionFile';
import { UserInfoProvider } from '@dc/hooks/useUserInfo';
import {
  ASSIGNMENT_SUBMISSION_TYPES,
  OPPORTUNITY_APPLICATION_STATUS,
  OPPORTUNITY_TYPE,
  VISIBILITY_SCOPE,
} from '@dc/resources/enums';
import { TAssignment } from '@dc/components/Student/Lesson/types';
import { VIRTUAL_INTERNSHIP_LESSON_QUERY } from '@dc/graphql/student/queries/virtualInternshipLesson';
import createAssignmentSubmissionMutation from '@dc/graphql/student/mutations/createAssignmentSubmission';
import { UPDATE_ASSIGNMENT_SUBMISSION_MUTATION } from '@dc/graphql/student/mutations/updateAssignmentSubmission';
import {
  ASSESSMENT_SUBMISSION_STATUS,
  ASSET_TYPE,
  LESSON_TYPES,
  RESOURCE_CLASS,
} from '@dc/resources/constants';
import generatePresignedUrlMutation from '@dc/graphql/student/mutations/generatePresignedUploadUrl';
import createAssignmentSubmissionFileMutation from '@dc/graphql/student/mutations/createAssignmentSubmissionFile';
import { studentInfoMock } from '@dc/tests/mocks/studentMocks';

import { NavigationContextProvider } from '@shared/components/Sidebar/useNavigation';

const opportunityFetchSpy = jest.fn();
const deleteAssignmentFetchSpy = jest.fn();
const createAssignmentFileSpy = jest.fn();
const updateAssignmentFetchSpy = jest.fn();
const generatePresignatedUrlFetchSpy = jest.fn();
const generateFilesFetchSpy = jest.fn();

const xhrMock = {
  onreadystatechange: null,
  open: jest.fn(),
  send: jest.fn(),
  upload: {},
};

const virtualInternshipLessonMock = {
  request: {
    query: VIRTUAL_INTERNSHIP_LESSON_QUERY,
    variables: { opportunityId: '14', lessonId: '4' },
  },
  result() {
    opportunityFetchSpy();

    return {
      data: {
        opportunity: {
          applicationStatus: OPPORTUNITY_APPLICATION_STATUS.PENDING,
          opportunityApplication: { id: '123' },
          archivedAt: '',
          automaticAcceptance: true,
          availableSpots: 3,
          creditsOutcomes: 'Credits',
          deadline: '',
          periodStart: '',
          periodEnd: '',
          tags: [],
          description: 'Description',
          id: '1',
          imageUrl: 'url-to-image',
          isFavorite: true,
          opportunityType: OPPORTUNITY_TYPE.VIRTUAL_INTERNSHIP,
          pathways: [{ id: '1', name: 'Pathway name' }],
          salaryInformation: 'Salary information',
          location: 'location',
          name: 'Test Virtual Internship',
          questions: [{ id: '1', question: 'Question?' }],
          visibilityScope: VISIBILITY_SCOPE.ENTITY,
          virtualInternship: {
            id: '1000',
            requiredExperiences: 2,
            lesson: {
              assignments: [],
              description: {
                audience: '',
                goal: '',
                introduction: '',
                role: '',
                situation: '',
              },
              progress: { performed: 1, total: 2, submitted: 0, accepted: 1 },
              attachments: [],
              vocabularies: [],
              careerReviewSurvey: { version: 1, questions: [], performed: false },
              checkInGroups: [],
              checkInQuestions: [],
              videos: [],
              type: LESSON_TYPES.GENERIC,
              texts: [],
              researchLinks: [],
              externalPresentations: [],
              name: 'Lesson name',
              imageUrl: 'image-url',
              id: '1',
              hasPresentation: true,
            },
            readinessSkillsLessons: [
              {
                id: '22',
                name: 'First readiness skills lessons name',
                thumbnailUrl: 'thumbnail-url',
                progress: { accepted: 0, submitted: 0, total: 0 },
              },
              {
                id: '33',
                name: 'Second readiness skills lessons name',
                thumbnailUrl: 'thumbnail-url',
                progress: { accepted: 0, submitted: 0, total: 0 },
              },
            ],
            experienceOpportunityLessons: [
              {
                id: '123',
                name: 'Preselected experience opportunity lesson name',
                thumbnailUrl: 'thumbnail-url',
                description: {
                  introduction: 'Experience opportunity description - id: 123',
                  audience: '',
                  goal: '',
                  role: '',
                  situation: '',
                },
                progress: { accepted: 0, submitted: 0, total: 0 },
              },
              {
                id: '124',
                name: 'Experience opportunity lesson to select',
                thumbnailUrl: 'thumbnail-url',
                description: {
                  introduction: 'Experience opportunity description - id: 124',
                  audience: '',
                  goal: '',
                  role: '',
                  situation: '',
                },
                progress: { accepted: 0, submitted: 0, total: 0 },
              },
            ],
            studentExperienceOpportunityLessons: [
              {
                id: '123',
                name: 'Preselected experience opportunity lesson name',
                thumbnailUrl: 'thumbnail-url',
                description: {
                  introduction: 'Experience opportunity description - id: 123',
                  audience: '',
                  goal: '',
                  role: '',
                  situation: '',
                },
                progress: { accepted: 0, submitted: 0, total: 0 },
              },
            ],
            postExperienceLessons: [
              {
                id: '11',
                name: 'First post experience lessons name',
                thumbnailUrl: 'thumbnail-url',
                progress: { accepted: 0, submitted: 0, total: 0 },
              },
              {
                id: '12',
                name: 'Second post experience lessons name',
                thumbnailUrl: 'thumbnail-url',
                progress: { accepted: 0, submitted: 0, total: 0 },
              },
            ],
            calendarLessons: [
              {
                id: '1',
                name: 'First lesson name',
                thumbnailUrl: 'thumbnail-url',
                progress: { accepted: 0, submitted: 0, total: 0 },
              },
              {
                id: '2',
                name: 'Second lesson name',
                thumbnailUrl: 'thumbnail-url',
                progress: { accepted: 0, submitted: 0, total: 0 },
              },
              {
                id: '3',
                name: 'Third lesson name',
                thumbnailUrl: 'thumbnail-url',
                progress: { accepted: 0, submitted: 0, total: 0 },
              },
              {
                id: '4',
                name: 'Fourth lesson name',
                thumbnailUrl: 'thumbnail-url',
                progress: { accepted: 0, submitted: 0, total: 0 },
              },
            ],
          },
        },
      },
    };
  },
};

const deleteAssignmentFileMock = {
  request: {
    query: deleteAssignmentSubmissionFileMutation,
    variables: { input: { id: '1' } },
  },
  result() {
    deleteAssignmentFetchSpy();

    return {
      data: {
        deleteAssignmentSubmissionFile: {
          status: 'ok',
        },
      },
    };
  },
};

const createAssigmentFileMock = {
  request: {
    query: createAssignmentSubmissionMutation,
    variables: {
      input: {
        assignmentId: '1',
        contextId: '1000',
        contextType: ASSIGNMENT_SUBMISSION_TYPES.VIRTUAL_INTERNSHIP,
      },
    },
  },
  result() {
    createAssignmentFileSpy();

    return {
      data: {
        createAssignmentSubmission: {
          assignmentSubmission: {
            id: '1',
          },
        },
      },
    };
  },
};
const updateAssignmentFileMock = {
  request: {
    query: UPDATE_ASSIGNMENT_SUBMISSION_MUTATION,
    variables: { input: { id: '1', status: ASSESSMENT_SUBMISSION_STATUS.SUBMITTED } },
  },
  result() {
    updateAssignmentFetchSpy();

    return {
      data: {
        updateAssignmentSubmission: {
          assignmentSubmission: {
            id: '1',
            status: ASSESSMENT_SUBMISSION_STATUS.SUBMITTED,
          },
        },
      },
    };
  },
};
const generatePresignitedUrlFileMock = {
  request: {
    query: generatePresignedUrlMutation,
    variables: {
      input: {
        filename: 'first_file.pdf',
        assetType: ASSET_TYPE.FILE,
        resourceClass: RESOURCE_CLASS.ASSIGNMENT_SUBMISSION_FILE,
      },
    },
  },
  result() {
    generatePresignatedUrlFetchSpy();

    return {
      data: {
        generatePresignedUploadUrl: {
          url: 'someurl',
          uuid: '1234-1234-1234',
        },
      },
    };
  },
};

const createAssignmentSubmissionFile = {
  request: {
    query: createAssignmentSubmissionFileMutation,
    variables: {
      input: {
        fileFilename: 'first_file.pdf',
        fileUuid: '1234-1234-1234',
        assignmentSubmissionId: '1',
      },
    },
  },
  result() {
    generateFilesFetchSpy();

    return {
      data: {
        assignmentSubmissionFile: {
          filename: 'first_file.pdf',
          id: '1234-1234-1234',
          url: 'someurl',
        },
      },
    };
  },
};

const opportunityMocks = [
  virtualInternshipLessonMock,
  deleteAssignmentFileMock,
  updateAssignmentFileMock,
  createAssigmentFileMock,
  generatePresignitedUrlFileMock,
  createAssignmentSubmissionFile,
  studentInfoMock,
];

const defaultAssignment = {
  id: '1',
  name: 'Some  name',
  description: 'Some description',
  displayName: 'Some display name',
  step: 2,
  rubrics: [],
  updatedAt: '2020-01-01T00:00:00.000Z',
  submission: {
    files: [],
    grade: null,
    rubricGrade: null,
    updatedAt: '2020-01-01T00:00:00.000Z',
    id: '1',
    status: 'DRAFT' as const,
  },
  __typename: 'Assignment' as const,
};

const assignmentWithFiles = {
  ...defaultAssignment,
  submission: {
    ...defaultAssignment.submission,
    files: [
      { id: '1', filename: 'somename1.pdf', source: '', url: 'file-url' },
      { id: '2', filename: 'somename2.pdf', source: '', url: 'file-url' },
      { id: '3', filename: 'somename3.pdf', source: '', url: 'file-url' },
      { id: '4', filename: 'somename4.pdf', source: '', url: 'file-url' },
    ],
  },
};

const assignmentWithoutFiles = {
  id: '1',
  name: 'Some  name',
  description: 'Some description',
  displayName: 'Some display name',
  step: 2,
  rubrics: [],
  updatedAt: '2020-01-01T00:00:00.000Z',
  submission: null,
  __typename: 'Assignment' as const,
};

const renderAssigment = (assignment: TAssignment = defaultAssignment) =>
  renderWithRouterAndReduxProvider(
    <MockedProvider mocks={opportunityMocks}>
      <UserInfoProvider
        value={{ userInfo: { ...studentInfoMock.result.data.userInfo, uuid: '2' } }}>
        <NavigationContextProvider>
          <Route path='/opportunities/:opportunityId/virtual-internship/lesson/:lessonId'>
            <AssignmentCard assignment={assignment} virtualInternshipId='1000' />
          </Route>
        </NavigationContextProvider>
      </UserInfoProvider>
    </MockedProvider>,
    { route: '/opportunities/14/virtual-internship/lesson/4' }
  );

describe('StudentVirtualInternshipAssignment', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    //@ts-ignore
    global.XMLHttpRequest = jest.fn(() => xhrMock);
  });

  it('renders correctly', async () => {
    const { container } = renderAssigment();
    expect(container).toMatchSnapshot();
  });

  it('renders empty files list when no files', async () => {
    renderAssigment();

    const assignmentList = await screen.queryAllByRole('listitem');
    const submitButton = await screen.findByRole('button', { name: 'Submit' });

    expect(submitButton).toBeDisabled();
    expect(assignmentList).toHaveLength(0);
  });

  it('renders files list with added files', async () => {
    renderAssigment(assignmentWithFiles);

    const assignmentList = await screen.queryAllByRole('listitem');
    const submitButton = await screen.findByRole('button', { name: 'Submit' });

    expect(submitButton).toBeEnabled();
    expect(assignmentList).toHaveLength(4);
  });

  it('should call delete file mutation when remove button clicked ', async () => {
    renderAssigment(assignmentWithFiles);

    const deleteButtons = await screen.findAllByRole('button', { name: 'Remove file' });
    userEvent.click(deleteButtons[0]);
    const dialog = await screen.findByRole('dialog');
    const archiveButton = within(dialog).getByRole('button', { name: 'Archive' });
    expect(dialog).toBeInTheDocument();

    userEvent.click(archiveButton);
    await waitFor(() => {
      expect(deleteAssignmentFetchSpy).toHaveBeenCalledTimes(1);
    });
  });

  it('should calls create submission', async () => {
    renderAssigment(assignmentWithoutFiles);

    const dropZone = await screen.findByTestId('drop-zone-input');
    fireEvent.change(dropZone, {
      target: {
        files: [{ name: 'first_file.pdf', size: 100 }],
      },
    });

    await waitFor(() => {
      expect(createAssignmentFileSpy).toHaveBeenCalledTimes(1);
    });
  });
});
