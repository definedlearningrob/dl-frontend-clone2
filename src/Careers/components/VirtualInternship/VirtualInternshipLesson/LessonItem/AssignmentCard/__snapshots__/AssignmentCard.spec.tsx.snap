// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StudentVirtualInternshipAssignment renders correctly 1`] = `
<div>
  <div
    class="card lessonCard"
    data-testid="lesson-element-card"
    id="1-Assignment"
  >
    <div
      class="cardHeader"
    >
      <div
        class="flex items-center gap-x xxxl:gap-sm"
      >
        <h5
          class="cardTitle"
        >
          Some display name
        </h5>
      </div>
      <div
        class="flex gap-xs xxxl:gap-sm items-center"
      >
        <button
          class="button flex items-center justify-center gap-xs rounded-sm cursor-pointer text-sm font-medium no-underline focus-visible:outline focus-visible:outline-1 focus-visible:outline-offset-[3px] -primary-outlined text-primary-500 hover:bg-primary-200 py-xs px-sm !text-xs leading-[14px] !gap-xxs border"
          data-testid="button"
          type="button"
        >
          <div
            class="flex"
          >
            <div
              class="flex align-center justify-center grow-0 shrink-0 h-[14px] mt-[-1px]"
              data-testid="icon"
            >
              <div
                class="h-sm w-sm"
              />
            </div>
          </div>
          <span
            class="items-center flex justify-center gap-xs"
          >
            Rubrics
          </span>
        </button>
        <span
          data-state="closed"
        >
          <button
            aria-label="Ask for guidance"
            class="flex items-center justify-center transition-colors rounded-sm border border-transparent cursor-pointer focus-visible:outline focus-visible:outline-1 focus-visible:outline-offset-[3px] !border-primary-500 text-primary-500 hover:bg-primary-200 outline-primary-500  w-md h-md"
            data-testid="icon-button"
            type="button"
          >
            <div
              class="flex"
            >
              <div
                class="flex align-center justify-center grow-0 shrink-0"
                data-testid="icon"
              >
                <div
                  class="h-base w-base"
                />
              </div>
            </div>
          </button>
        </span>
      </div>
    </div>
    <p
      class="assignmentDescription"
    >
      Some description
    </p>
    <div
      class="filesUploadSection"
    >
      <div
        class="dropzoneWrapper"
      >
        <div
          class="wrapper"
        >
          <header
            class="header"
          >
            <button
              class="wrapper selected"
              type="button"
            >
              <div
                class="icon icon -sm-size"
                data-testid="icon"
              >
                <div />
              </div>
              <span
                class="text"
              >
                File
              </span>
            </button>
            <button
              class="wrapper"
              type="button"
            >
              <div
                class="icon icon -sm-size"
                data-testid="icon"
              >
                <div />
              </div>
              <span
                class="text"
              >
                Google Drive
              </span>
            </button>
          </header>
          <div
            class="bg-center bg-cover bg-white flex items-center justify-center group rounded-sm cursor-pointer border-2 border-dashed border-neutral-300 hover:bg-neutral-200 hover:border-primary-500 relative p-0 pt-[50%] text-center text-xs xxxl:text-sm !pt-0 h-[120px]"
            data-testid="drop-container"
            tabindex="0"
          >
            <div
              class="absolute inset-0"
            >
              <input
                autocomplete="off"
                data-testid="drop-zone-input"
                id="drop-zone-input-undefined"
                multiple=""
                style="display: none;"
                tabindex="-1"
                type="file"
              />
              <div
                class="flex flex-col justify-center items-center h-full p-sm"
              >
                <div
                  class="flex"
                >
                  <div
                    class="flex align-center justify-center grow-0 shrink-0 p-xs mb-xs text-primary-500 bg-primary-200 rounded-full group-hover:bg-white p-sm"
                    data-testid="icon"
                  >
                    <div
                      class="h-base w-base"
                    />
                  </div>
                </div>
                <p
                  class="text-primary-500 m-0 z-lowest"
                >
                  Drop or 
                  <span
                    class="text-primary-500 underline"
                  >
                    choose file
                  </span>
                   to upload
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="submissionFooter"
    >
      <div
        class="submissionStatus"
      >
        <div
          class="kicker md default"
        >
          Assignment status:
        </div>
        <div
          class="submissionStatusValue"
        >
          <div
            class="font-medium leading-lg flex items-center whitespace-nowrap bg-secondary-200 text-secondary-500 rounded-sm p-xxs text-xs gap-xxs"
          >
            <div
              class="flex"
            >
              <div
                class="flex align-center justify-center grow-0 shrink-0"
                data-testid="icon"
              >
                <div
                  class="h-sm w-sm"
                />
              </div>
            </div>
            Draft
          </div>
        </div>
      </div>
      <button
        class="button flex items-center justify-center gap-xs rounded-sm cursor-pointer text-sm font-medium no-underline focus-visible:outline focus-visible:outline-1 focus-visible:outline-offset-[3px] -primary bg-primary-500 hover:bg-primary-600 outline-primary-500 border-primary-500 py-xs px-base leading-[22px] border text-white"
        data-testid="status-submit-button"
        disabled=""
        type="button"
      >
        <span
          class="items-center flex justify-center gap-xs"
        >
          Submit
        </span>
      </button>
    </div>
  </div>
</div>
`;
